provider/azure/metadata.go

```diff
@@ -0,0 +1,119 @@
+// This package reads from the Azure metadata service; see
+// https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service#custom-data
+package azure
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/taskcluster/httpbackoff/v3"
+)
+
+var MetadataBaseURL = "http://169.254.169.254"
+
+// Data from the `/instance` endpoint.  Note that this is a partial set --
+// feel free to add additional fields here as necessary.
+type InstanceData struct {
+	Compute struct {
+		CustomData string `json:"customData"`
+		Location   string `json:"location"`
+		VMID       string `json:"vmId"`
+		VMSize     string `json:"vmSize"`
+	} `json:"compute"`
+	Network struct {
+		Interface []struct {
+			IPV4 struct {
+				IPAddress []struct {
+					PrivateIPAddress string `json:"privateIpAddress"`
+					PublicIPAddress  string `json:"publicIpAddress"`
+				} `json:"ipAddress"`
+			} `json:"ipv4"`
+		} `json:"interface"`
+	} `json:"network"`
+}
+
+// Data from the /scheduledevents endpoint
+type ScheduledEvents struct {
+	Events []struct {
+		EventId      string
+		EventType    string
+		ResourceType string
+		Resources    []string
+		EventStatus  string
+		NotBefore    string
+	}
+}
+
+// Data from the /attested/document endpoint.
+type AttestedDocument struct {
+	Encoding  string `json:"encoding"`
+	Signature string `json:"signature"`
+}
+
+type MetadataService interface {
+	// Query the /instance endpoint
+	queryInstanceData() (*InstanceData, error)
+	// Get the content of the Attested document
+	queryAttestedDocument() (string, error)
+	// Get the content of the scheduled events
+	queryScheduledEvents() (*ScheduledEvents, error)
+}
+
+type realMetadataService struct{}
+
+func (mds *realMetadataService) fetch(path string, apiVersion string) (string, error) {
+	client := http.Client{}
+	req, err := http.NewRequest("GET", MetadataBaseURL+path+"?api-version="+apiVersion, nil)
```

Comment:
At the point where this is this complex would it make more sense to start using `url.URL`?

Reply:
The result is quite a bit longer and less clear, mostly because the API for Query isn't very natural:

```diff
@@ -60,17 +61,22 @@ type MetadataService interface {
        // Get the content of the scheduled events
        queryScheduledEvents() (*ScheduledEvents, error)
 }
 
 type realMetadataService struct{}
 
 func (mds *realMetadataService) fetch(path string, apiVersion string) (string, error) {
        client := http.Client{}
-       req, err := http.NewRequest("GET", MetadataBaseURL+path+"?api-version="+apiVersion, nil)
+       url, _ := url.Parse(MetadataBaseURL)
+       url.Path = path
+       query := url.Query()
+       query.Add("api-version", apiVersion)
+       url.RawQuery = query.Encode()
+       req, err := http.NewRequest("GET", url.String(), nil)
        if err != nil {
                return "", err
        }
 
        req.Header.Set("Metadata", "true")
 
        resp, _, err := httpbackoff.ClientDo(&client, req)
        if err != nil {
```

I think the string concatenation form is more readble..

Reply:
Ah, based on Slack I made a shorter version.